luigi/parameter.py

```diff
@@ -1117,7 +1139,12 @@ def normalize(self, value):
         """
         frozen_value = recursively_freeze(value)
         if self.schema is not None:
-            jsonschema.validate(instance=recursively_unfreeze(frozen_value), schema=self.schema)
+            unfrozen_value = recursively_unfreeze(frozen_value)
+            try:
+                self.schema.validate(unfrozen_value)  # Validators may update the instance inplace
+                frozen_value = super().normalize(unfrozen_value)
```

Comment:
I understand the try/except here to handle for when `self.schema` is a custom validator, but could you explain why you re-freeze the unfrozen value again? Why/how would the validator modify the value of the instance such that we cannot return the `frozen_value` from the beginning of this method?

Reply:
It is possible to define custom validators to handle specific schemas or do extra stuffs: https://python-jsonschema.readthedocs.io/en/stable/creating/
For example, it is possible to fill missing values with default values: https://python-jsonschema.readthedocs.io/en/stable/faq/#why-doesn-t-my-schema-s-default-property-set-the-default-on-my-instance
In this example the given instance is updated with default values during the validation, so the new instance must be frozen again and returned.
Nevertheless, we can indeed simplify this a bit: unfreeze the given value when a schema is provided and just freeze the result, instead of freezing, unfreezing and freezing again. I pushed a new commit for this.